### ISSUE TITLE
Remove Javax.* dependency from library code

### ISSUE THREAD

**Issue by olidroide** [immutables/immutables#142]
When I try to compile with 2.0.16 version 
```
InvalidPackage: Package not included in Android
../../../../../../.gradle/caches/modules-2/files-2.1/org.immutables/gson/2.0.16/b931027465928941c506c82347c325806bed8ad2/gson-2.0.16.jar:
Invalid package reference in library; not included in Android: javax.ws.rs.core. Referenced from org.immutables.gson.stream.GsonMessageBodyProvider.
../../../../../../.gradle/caches/modules-2/files-2.1/org.immutables/gson/2.0.16/b931027465928941c506c82347c325806bed8ad2/gson-2.0.16.jar:
Invalid package reference in library; not included in Android: javax.ws.rs. Referenced from org.immutables.gson.stream.GsonMessageBodyProvider.
../../../../../../.gradle/caches/modules-2/files-2.1/org.immutables/value/2.0.16/c9538226160a1377028e5bb7922a0ca1d409d92/value-2.0.16.jar:
Invalid package reference in library; not included in Android: javax.tools. Referenced from org.immutables.value.internal..generator...Output.AppendServiceFile.
Priority: 6 / 10
Category: Correctness
Severity: Error
Explanation: Package not included in Android.
```

Is possible to remove these dependencies with ```javax```

**Comment by elucash** [comment]
Thank you for the issue report!
Those jars should be compile-only dependencies. To be more specific, `org.immutables:value` should be used only for compilation, and `org.immutables:gson` artifact could be used either at compile time or compile and runtime for java (if you need optional runtime features such as `javax.ws.rs` integration which is not applicable to Android anyway).

Using recent version of Android Studio and Gradle I was able to just use `provided` scope for  dependencies, this scope is not a built-in scope and is provided by `com.android.tools.build:gradle` plugins AFAIK.

```
dependencies {
    compile fileTree(dir: 'libs', include: ['*.jar'])
    provided "org.immutables:value:2.0.16"
    provided "org.immutables:gson:2.0.16"
}
```

There are other gradle-specific tricks how to make it work. Let us know if it will resolve the issue, otherwise we could look for other workarounds (or make changes to artifact structure if deemed absolutely necessary)

**Comment by elucash** [comment]
JFYI There's newly added section in "get started" page describing how to configure Immutables for Android builds: http://immutables.github.io/getstarted.html#android

**Comment by olidroide** [comment]
Nice! thanks!

**Comment by olidroide** [comment]
@elucash I create a small example repo https://github.com/olidroide/Immutables-Android-Example

**Comment by elucash** [comment]
Thank you for exhaustive example of the tools/setup you use, this helped a lot! So while Immutables are doing nothing wrong here, I would suspect the linter tool which do not differentiate `provided` from `compile` scoped dependencies. But we don't want you to fight tools, for such cases in current snapshot versions `2.1.0-SNAPSHOT` we introduced annotation-only version of jars with classifier `annotations` for `gson` and `value` artifacts

```javascript
repositories {
        ...
        maven {
           // You can get snapshots from sonatype oss repo
           //  this repository is not needed for release artifacts
            url "https://oss.sonatype.org/content/repositories/snapshots"
        }
}
```

```javascript
ext {
    supportLibrary = '22.2.1'
    immutablesLibrary = '2.1.0-SNAPSHOT'

    //Support Libraries dependencies
    appDependencies = [
            design                  : "com.android.support:design:$supportLibrary",
            recyclerView            : "com.android.support:recyclerview-v7:$supportLibrary",
            cardView                : "com.android.support:cardview-v7:$supportLibrary",
            appCompat               : "com.android.support:appcompat-v7:$supportLibrary",
            immutablesValue         : "org.immutables:value:$immutablesLibrary",

            immutablesValueAn         : "org.immutables:value:$immutablesLibrary:annotations", // <-- classifier 'annotations'
            immutablesBuilder       : "org.immutables:builder:$immutablesLibrary",
            immutablesGson          : "org.immutables:gson:$immutablesLibrary:annotations",  // <-- classifier 'annotations'
    ]
}

 ...
dependencies {
    ...
    apt appDependencies.immutablesValue
    provided appDependencies.immutablesValueAn
    provided appDependencies.immutablesBuilder
    provided appDependencies.immutablesGson
}
```

In this way you will will not see linter's complaints. Release of 2.1 is expected next month, I hope you have enough workarounds in a meantime (like ignoring lint warnings or using snapshot dependencies).
I'll update documentation (get started page) to mention this when release will be ready.

**Comment by olidroide** [comment]
great! thank you :+1: I will try this workaround, and update the example with the 2.1 library when it is released. I'm happy to be useful with this small example of Immutables.

**Comment by elucash** [comment]
JFYI this was released in 2.0.19, so no need to wait till 2.1

**Comment by olidroide** [comment]
Thanks! I already update my example ;)
